From 29b33ca94b2cccc52a3e0f94e435ed507f0b75b3 Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Tue, 25 Nov 2014 16:36:19 -0500
Subject: [fs] overlayfs: don't poison cursor

Message-id: <20141125163619.18115.41768.stgit@warthog.procyon.org.uk>
Patchwork-id: 100945
O-Subject: [PATCH 45/47] ovl: don't poison cursor
Bugzilla: 985875
RH-Acked-by: Steve Best <sbest@redhat.com>

From: Miklos Szeredi <miklos@szeredi.hu>

ovl_cache_put() can be called from ovl_dir_reset() if the cache needs to be
rebuilt.  We did list_del() on the cursor, which results in an Oops on the
poisoned pointer in ovl_seek_cursor().

Reported-by: Jordi Pujol Palomer <jordipujolp@gmail.com>
Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
Tested-by: Jordi Pujol Palomer <jordipujolp@gmail.com>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Original-commit: 3f822c6264954660babce757fb45792fd3af273e
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/overlayfs/readdir.c b/fs/overlayfs/readdir.c
index 4046b45..c003330 100644
--- a/fs/overlayfs/readdir.c
+++ b/fs/overlayfs/readdir.c
@@ -173,7 +173,7 @@ static void ovl_cache_put(struct ovl_dir_file *od, struct dentry *dentry)
 {
  struct ovl_dir_cache *cache = od->cache;
 
- list_del(&od->cursor.l_node);
+ list_del_init(&od->cursor.l_node);
  WARN_ON(cache->refcount <= 0);
  cache->refcount--;
  if (!cache->refcount) {
-- 
1.7.1